lib/mongo/server/connection_pool.rb

```diff
@@ -523,30 +541,70 @@ def populate
 
         catch(:done) do
           loop do
+            connection = nil
             @lock.synchronize do
               if !closed? && unsynchronized_size < min_size
-                @available_connections << create_connection
+                connection = create_connection
+                @pending_connections << connection
               else
                 throw(:done)
               end
             end
+
+            begin
+              authenticate_connection(connection)
+              @lock.synchronize do
+                if !closed?
+                  # TODO need to check it still exists in pending here,
+                  # in case pool is closed before we re-obtain the lock
+                  @available_connections << connection
+                  @pending_connections.delete(connection)
+
+                  # wake up one thread waiting for connections, since one was created
+                  @available_semaphore.signal
+                end
+              end
+            rescue Exception => e
+              @lock.synchronize do
+                @pending_connections.delete(connection)
```

Comment:
Closing the connection here is probably a good idea, in case the exception happened during handshake/auth after the socket got connected.